### DISCUSSION TITLE
LAG and Switches with Modules

### DISCUSSION THREAD

**Discussion by SteffenBernd** [netbox-community/netbox#7580] [Q&A]
Hello,
i have a Switch with 6 Modules.
I created the Switch as a Device with 6 Bays and every Module as a child Device.
I sloted the Child Module Devices into the Bays and would like to Create a LAG across different Modules.

Sadly this is not posibble. I cannot see the LAG on the Child Devices.
Is there a Workarround or a Fix for that?

Greetings
Steffen

**Comment by PieterL75** [comment]
Steffen,

I had the same issue #7355 
Seems that I was using the concept of Bays and Inventory wrong
Brian gave a good explanation on it in that post.

In short : never use Bays for switches and modules. Bays are used for a chassis that contains servers.
For switches, you have to create one device, with all ports on that one device, and in the inventory you can specify the details of each slot

Pieter

**Comment by SteffenBernd** [comment]
Hello,

it has been a long time and a few updates. is there a new take on this situation?
I havend found anything to solve my problem :-/

Greetings
Steffen

**Reply by candlerb** [reply]
Yes - Netbox [3.2](https://github.com/netbox-community/netbox/releases/tag/v3.2.0) includes a new model of "Modules" and "Module Bays" which do exactly what you want.

**Reply by SteffenBernd** [reply]
ok could you please explain where i can find it? and can i edit my existing stuff so i can use it?
i dont want to remake every switch we have. thats a lot of  work with the cabeling and so on.

**Reply by candlerb** [reply]
> could you please explain where i can find it?

The feature is described in issue #7844 linked from the v3.2.0 release notes, and in the documentation [here](https://docs.netbox.dev/en/stable/core-functionality/modules/).  There are submenus for "Modules" and "Module Types" under "Devices":

![image](https://user-images.githubusercontent.com/44789/176849913-4caab894-d83a-4538-b2d4-f9751d5c5d22.png)

And module bays are added to a device as "Components", in the same way as interfaces, power ports and so on.

> and can i edit my existing stuff so i can use it?
> i dont want to remake every switch we have. thats a lot of work with the cabeling and so on.

Yeah, that could be tricky.  You can modify your Device Types to include module bays, and you can retro-fit those module bays to existing instances of that type using [this script](https://github.com/netbox-community/reports/blob/master/scripts/add_device_type_components.py). That will allow you to use the feature going forwards for new line cards.

However, moving existing interfaces which are attached directly to the device, to be interfaces on modules, is trickier - especially because Netbox does not allow cables to be reconnected at one end, they must be deleted and recreated.

The only approach I know for this (which I've used to bulk-move cables from one switch to a different one) is:
1. Export the relevant cables as CSV. This requires a custom export template(*), because Netbox's own CSV export and import formats are not compatible with each other
2. Delete the old cables
3. Modify the cables CSV to set the new correct endpoints as desired
4. Import the CSV

In your case you'd need a different step 3:

3a. delete the old interfaces from the device
3b. delete the old inventory items (which represented line cards) from the device
3c. add the line card modules - which should create new interfaces

If the new interfaces on the modules have the same names as the original ones, then you won't need to modify the CSV data itself before re-importing.

--------
(*) This is the export template I used, for content type `dcim | cable`:

```
side_a_device,side_a_type,side_a_name,side_b_device,side_b_type,side_b_name,type,status,label,color,length,length_unit
{%- for c in queryset %}
{%- if c.termination_a_type.app_label != 'circuits' and c.termination_b_type.app_label != 'circuits' %}
{{ c.termination_a.device.name }},{{ c.termination_a_type.app_labeled_name|replace(" | ",".")|replace(" ","") }},{{ c.termination_a.name }},{{ c.termination_b.device.name }},{{ c.termination_b_type.app_labeled_name|replace(" | ",".")|replace(" ","") }},{{ c.termination_b.name }},{{ c.type }},{{ c.status }},"{{ c.label }}",{{c.color}},{{ c.length|replace("None","") }},{{ c.length_unit }}
{%- endif %}
{%- endfor %}
```

**Reply by candlerb** [reply]
Sorry, please ignore all that, I'm talking rubbish.

I've just tested it.  Not only are module interfaces instantiated directly as device interfaces; there's also the option when adding a module to a device to "adopt" existing interfaces of the same name.

![image](https://user-images.githubusercontent.com/44789/176852550-1ff61a94-6e9f-4e9d-bda4-40c482c9a2ee.png)

So you should be fine: no messing around with recreating cables is required.

Make sure you use `{module}` in your interface name templates attached to the module type. This is replaced with the "position" value from the corresponding module bay on the device, when instantiating a module.

**Reply by SteffenBernd** [reply]
hey thanks for the info :)
but im kinda lost right now. could you start from the beginning on what i have to do?
i have a divice with device bays, where i have devices sloted that have the interfaces.

What do i need to do step by step to create a module bay, module and copy everything from a device (wich is is sloted in the device bay) to that new modul?

**Reply by candlerb** [reply]
Firstly, I suggest you should set up a test instance of Netbox to learn the feature, or try it on https://demo.netbox.dev/, before doing anything on your live system.

However, if you've used *device bays* instead of *inventory items* for your line cards, then you've done it completely wrong from the start. You were warned clearly in the documentation not to do this ([here](https://docs.netbox.dev/en/stable/core-functionality/devices/#device-bays) and [here](https://docs.netbox.dev/en/stable/core-functionality/device-types/#device-types_1)).  Device bays hold independent devices - they are essentially physical subdivisions of a rack.  They do not model components of a single device.  A router with 5 line cards is not 6 independent devices, it's 1 device.

Hence if you've used device bays, then I'm afraid it will be a lot more work.  You will need to export all the cables connected to those child "devices"; delete all the child devices and device bays; add module bays to the devices; add the modules; and then re-import the cables. I included the export template you need above. You definitely want to take a database snapshot first, and try it out on a test system.

**Reply by SteffenBernd** [reply]
ok thank, i´ll look into it :)
thanks again for all your help :)

**Comment by SteffenBernd** [comment]
ok i´ve looked into it and its just inpractical...
the way i did it is just so mutch superiour to the modul method that i´ll keep it that way and type the Trunk information in the coments of the device i use as a modul.

**Reply by candlerb** [reply]
> the way i did it is just so mutch superiour to the modul method

Can you explain in what way - what's missing from the "module" approach that you'd like to see?

I can give you two immediate advantages of switching to modules for line cards:
1. When you look at a Device, under the Interfaces tab you will immediately see all the interfaces on the device, and what they are connected to.  You don't have to drill down to devices in device bays.
2. You can create a LAG between interfaces on different line cards

In addition: if you are using Netbox as a configuration source-of-truth, it's much better to see the interfaces on the device itself (which matches reality), rather than belonging to different devices.  This is particularly important for routing devices, where all the interface IPs really belong to a single device.

**Reply by SteffenBernd** [reply]
i didnt see that you can see the interfaces in the device. so that would be something good. but all modules have the same interface port names. its eth 1-x and i cannot do that. it always tells me that the interfaces already exist.

**Reply by candlerb** [reply]
> but all modules have the same interface port names. its eth 1-x and i cannot do that. it always tells me that the interfaces already exist

You've overlooked the `{module}` placeholder.

Name the interfaces on the module type as `GigabitEthernet{module}/0/[1-24]`

Then when you insert a module into a module bay on a device, the `{module}` string will be replaced by the "position" string from the module bay.  So if your have four module bays with positions 1,2,3,4 and you insert it into slot 4, then you'll get `GigabitEthernet4/0/1`, `GigabitEthernet4/0/2` etc.

This is described in the documentation for modules, but not in the help text shown on screen.  I've raised #9642 for this.

**Reply by SteffenBernd** [reply]
i´ll check this out after the weekend. thank you :)
